overlord/snapstate/snapstate.go

```diff
@@ -32,65 +32,49 @@ import (
 // Install returns a set of tasks for installing snap.
 // Note that the state must be locked by the caller.
 func Install(s *state.State, snap, channel string, flags snappy.InstallFlags) (*state.TaskSet, error) {
-	inst := installState{
+	// download
+	var download *state.Task
+	dl := downloadState{
 		Name:    snap,
 		Channel: channel,
 		Flags:   flags,
 	}
-
-	// download (if needed)
-	var download *state.Task
 	if !osutil.FileExists(snap) {
 		download = s.NewTask("download-snap", fmt.Sprintf(i18n.G("Downloading %q"), snap))
-		inst.DownloadTaskID = download.ID()
 	} else {
 		download = s.NewTask("nop", "")
-		inst.SnapPath = snap
+		dl.SnapPath = snap
 	}
-	download.Set("install-state", inst)
-
-	// check
-	check := s.NewTask("check-snap", fmt.Sprintf(i18n.G("Checking %q"), snap))
-	check.Set("install-state", inst)
-	check.WaitFor(download)
+	download.Set("download-state", dl)
```

Comment:
As we discussed, can we create the mount-state ahead of time here?

If so, I suggest adding SnapPath into mount-state itself, so we can drop the link between mount-snap and download-snap. Then, let's rename mount-snap-id to mount-state-id, so everybody else is looking at that same data and we can put that state in another task if we want to without it feeling awkward.